cluster-controller/src/main/java/io/strimzi/controller/cluster/resources/TopicController.java

```diff
@@ -0,0 +1,245 @@
+     * @param dep the deployment from which to recover the topic controller state
+     * @return Topic Controller instance, null if the corresponding Deployment doesn't exist
+     */
+    public static TopicController fromDeployment(String namespace, String cluster, Deployment dep) {
+
+        TopicController topicController = null;
+
+        if (dep != null) {
+
+            topicController = new TopicController(namespace, cluster);
+
+            topicController.setLabels(dep.getMetadata().getLabels());
+            topicController.setReplicas(dep.getSpec().getReplicas());
+            topicController.setImage(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getImage());
+            topicController.setHealthCheckInitialDelay(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getReadinessProbe().getInitialDelaySeconds());
+            topicController.setHealthCheckTimeout(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getReadinessProbe().getTimeoutSeconds());
+
+            Map<String, String> vars = dep.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().stream().collect(
+                    Collectors.toMap(EnvVar::getName, EnvVar::getValue));
+
+            TopicControllerConfig config = new TopicControllerConfig();
+
+            config.withImage(topicController.getImage())
+                    .withKafkaBootstrapServers(vars.getOrDefault(KEY_KAFKA_BOOTSTRAP_SERVERS, defaultBootstrapServers(cluster)))
+                    .withZookeeperConnect(vars.getOrDefault(KEY_ZOOKEEPER_CONNECT, defaultZookeeperConnect(cluster)))
+                    .withNamespace(vars.getOrDefault(KEY_NAMESPACE, namespace))
+                    .withReconciliationInterval(vars.getOrDefault(KEY_FULL_RECONCILIATION_INTERVAL, DEFAULT_FULL_RECONCILIATION_INTERVAL))
+                    .withZookeeperSessionTimeout(vars.getOrDefault(KEY_ZOOKEEPER_SESSION_TIMEOUT, DEFAULT_ZOOKEEPER_SESSION_TIMEOUT));
+
+            topicController.setConfig(config);
+        }
+
+        return topicController;
+    }
+
+    /**
+     * Return the differences between the current Topic Controller and the deployed one
+     *
+     * @param dep Deployment which should be diffed
+     * @return  ClusterDiffResult instance with differences
+     */
+    public ClusterDiffResult diff(Deployment dep) {
+
+        if (dep != null) {
+            String image = dep.getSpec().getTemplate().getSpec().getContainers().get(0).getImage();
+
+            Map<String, String> vars = dep.getSpec().getTemplate().getSpec().getContainers().get(0).getEnv().stream().collect(
+                    Collectors.toMap(EnvVar::getName, EnvVar::getValue));
+
+            // get the current (deployed) topic controller configuration
+            TopicControllerConfig depConfig = new TopicControllerConfig();
+
+            depConfig.withKafkaBootstrapServers(vars.getOrDefault(KEY_KAFKA_BOOTSTRAP_SERVERS, defaultBootstrapServers(cluster)))
```

Comment:
For my money, all these `with*()` methods are a bit point less. As an API it's easy to miss off setting one of the properties. And it's not like we use the mutability at all. I would just make all these properties parameters of the constructor. I know you're aiming for consistency with the other Cluster resources, but imho those make the same mistake and we should avoid perpetuating that mistake here.

Reply:
I was inspired by following the builder pattern but in this case without using a real builder just having a fluent API on the final class to build. Does it work if I use the same way you used for `Topic` and `Topic.Builder` in the topic controller ? It should be the same, or ?

Reply:
With `Topic` there's the need to make a changed copy of a `Topic`, but `Topic` is immutable. If we only ever changed a single property of a `Topic` then a builder still wouldn't really justified, you'd just have method which returned a new one. I would argue that the Builder is justified because we sometimes change several or all the propertes, and it would be silly to create many intermediate `Topics` to do that without a `Builder`. (The reason `Topic` is immutable is they're freely shared between threads, so it's safer).

I don't think that's really the case here, where we only really create these things in one place, and have no need to create changed versions. So no need for a builder in my opinion, and if we use constructor parameters we can make this thing immutable too.

Reply:
When in the `Storage` class the `fromJson` method builds the instance reading fields from the JSON string it seems to me more natural to set the fields one by one than reading all the fields, saving in variables and then using a constructor with all parameters. The `with*` methods are just `set*` providing a fluent API returning the current instance which sometimes makes the code more readable (at least to me). For sure I would not say "make the same mistake and we should avoid perpetuating that mistake here" ... it doesn't seem to be a mistake to me, it's just a different point of view.